processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,45 +132,104 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
+    final Integer priorIdex = facts.get(key);
+
+    Aggregator[] aggs;
+
+    if (null != priorIdex) {
+      aggs = atomicGet(priorIdex);
+    } else {
+      aggs = new Aggregator[metrics.length];
+      for (int i = 0; i < metrics.length; i++) {
+        final AggregatorFactory agg = metrics[i];
+        aggs[i] = agg.factorize(
+            makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
+        );
+      }
+      final Integer rowIndex = indexIncrement.getAndIncrement();
+
+      // We always want to have a NULL value to replace
+      if (aggList.length <= rowIndex) {
+        ensureCapacity(aggList.length + aggListNullStepSize);
+      }
+
+      atomicSet(rowIndex, aggs);
+
+      // Last ditch sanity checks
+      if (numEntries.get() >= maxRowCount && !facts.containsKey(key)) {
         throw new IndexSizeExceededException("Maximum number of rows reached");
       }
-      final Integer prev = facts.putIfAbsent(key, rowOffset);
-      if (prev != null) {
-        rowOffset = prev;
-      } else {
-        Aggregator[] aggs = new Aggregator[metrics.length];
-        for (int i = 0; i < metrics.length; i++) {
-          final AggregatorFactory agg = metrics[i];
-          aggs[i] = agg.factorize(
-              makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
-          );
-        }
-        aggList.add(aggs);
+      final Integer prev = facts.putIfAbsent(key, rowIndex);
+      if (null == prev) {
         numEntries.incrementAndGet();
+      } else {
+        // We lost a race
+        aggs = atomicGet(prev);
+        // Free up the misfire
+        atomicSet(rowIndex, null);
+        // This is expected to occur ~80% of the time in the worst scenarios
       }
     }
 
     in.set(row);
 
-    final Aggregator[] aggs = aggList.get(rowOffset);
-    for (int i = 0; i < aggs.length; i++) {
-      synchronized (aggs[i]) {
-        aggs[i].aggregate();
+    for (Aggregator agg : aggs) {
+      synchronized (agg) {
+        agg.aggregate();
       }
     }
+
     in.set(null);
+
+
     return numEntries.get();
   }
 
+  private static final boolean wasCopying(int val){
+    return (val&1)>0;
+  }
+
+  private Aggregator[] atomicGet(int offset){
```

Comment:
this implementation is currently assuming that `volatile` on aggList is good enough. I can change it to AtomicReference to be absolutely sure.

Remember, these get and set are not on the aggList, but on an aggregator at an offset IN the aggList.